libc/sock/syslog.c

```diff
@@ -0,0 +1,325 @@
+         (e==EPIPE);
+}
+
+static void __openlog() {
+  if (IsWindows()) {
+    log_fd = RegisterEventSourceA(NULL, log_ident);
+  } else {
+    log_fd = socket(AF_UNIX, SOCK_DGRAM|SOCK_CLOEXEC, 0);
+    if (log_fd >= 0) connect(log_fd, (void *)&log_addr, sizeof(log_addr));
+  }
+}
+
+
+/**
+ * Generates a log message which will be distributed by syslogd
+ *
+ * Note: no errors are reported if an error occurred while logging
+ *       the message.
+ *
+ * @param priority is a bitmask containing the facility value and
+ *        the level value. If no facility value is ORed into priority,
+ *        then the default value set by openlog() is used.
+ *        it set to NULL, the program name is used.
+ *        Level is one of LOG_EMERG, LOG_ALERT, LOG_CRIT, LOG_ERR, 
+ *        LOG_WARNING, LOG_NOTICE, LOG_INFO, LOG_DEBUG
+ * @param message the format of the message to be processed by vprintf()
+ * @param ap the va_list of arguments to be applied to the message
+ * @asyncsignalsafe
+ */
+void vsyslog(int priority, const char *message, va_list ap) {
+  char timebuf[16];     /* Store formatted time */
+  time_t now;
+  struct tm tm;
+  char buf[1024];
+  int errno_save = errno; /* Save current errno */
+  int pid;
+  int l, l2;
+  int hlen; /* If LOG_CONS is specified, use to store the point in
+             * the header message after the timestamp */
+
+  if (log_fd < 0) __openlog();
+
+  if (!(priority & LOG_FACMASK)) priority |= log_facility;
+
+  /* Build the time string */
+  now = time(NULL);
+  gmtime_r(&now, &tm);
+  strftime(timebuf, sizeof(timebuf), "%b %e %T", &tm);
+
+  pid = (log_opt & LOG_PID) ? getpid() : 0;
+  /* This is a clever trick to optionally include "[<pid>]"
+   * only if pid != 0. When pid==0, the while "[%.0d]" is skipped:
+   *   %s%.0d%s -> String, pidValue, String
+   * Each of those %s:
+   *  - if pid == 0 -> !pid is true (1), so "[" + 1 points to the 
+   *    NULL terminator after the "[".
+   *  - if pid != 0 -> !pid is false (0), so the string printed is
+   *    the "[".
+   */
+  l = snprintf(buf, sizeof(buf), "<%d>%s ", priority, timebuf);
+  hlen = l;
+  l += snprintf(buf+l, sizeof(buf)-l, "%s%s%.0d%s: ", log_ident, "["+!pid, pid, "]"+!pid);
+  errno = errno_save;
+
+  /* Append user message */
+  l2 = vsnprintf(buf+l, sizeof(buf)-l, message, ap);
+  if (l2 >= 0) {
+    if (l2 >= sizeof(buf) - l) {
+      l = sizeof(buf) - 1;
+    } else {
+      l += l2;
+    }
+    if (buf[l-1] != '\n') {
+      buf[l++] = '\n';
+    }
+    if (!IsWindows()) {
+      /* Unix: 
+       * - First try to send it to syslogd
+       * - If fails and LOG_CONS is provided, writes to /dev/console
+       */
+      // TODO: If we use send() it will end up calling the send stubs under
+      //       sysv/calls/send.S and that will report ENOSYS (apparently that 
+      //       function serves only BSD). If we use sendto() it works as expected.
+      //       Why send() does not call the send() function defined in sock/send.c ?
```

Comment:
Wait, are you saying that `send(sockfd, buf, len, flags);` isn't equivalent to `sendto(sockfd, buf, len, flags, NULL, 0);` on BSD?

Reply:
No no... they are equivalent, of course. My problem was that if I was using `send()`, the linker would pick up the wrong BSD `send()` and not the correct one under `libc/sock/send.c`. Now that I have commented out that wrong definition, I changed the code to use `send()`. I have just rebuilt (I realized I needed to manually delete the `send.s` file and perform a clean build) and run my sendlog test. Now it works.